conf/imx_hab4/sign-spl.sh

```diff
@@ -0,0 +1,140 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+CST_BINARY="cst"
+CSF_TEMPLATE="u-boot-spl-sign.csf-template"
+SPL_FILE="SPL"
+KEY_DIR="."
+DCD_CLEAR=0
+DISPLAY_USAGE=0
+
+parse_args()
+{
+    while [ ${#} -gt 0 ]
+    do
+        case ${1} in
+        --cst)
+            CST_BINARY=${2}
+            shift
+            shift
+            ;;
+        --csf-template)
+            CSF_TEMPLATE=${2}
+            shift
+            shift
+            ;;
+        --spl)
+            SPL_FILE=${2}
+            shift
+            shift
+            ;;
+        --key-dir)
+            KEY_DIR=${2}
+            shift
+            shift
+            ;;
+        --mfgtools)
+            DCD_CLEAR=1
+            shift
+            ;;
+        -h)
+            DISPLAY_USAGE=1
+            shift
+            ;;
+        --help)
+            DISPLAY_USAGE=1
+            shift
+            ;;
+        *)
+            shift
+            ;;
+        esac
+    done
+}
+
+parse_args "$@"
+
+if [ "${DISPLAY_USAGE}" = "1" ]; then
+    echo "usage for: ${0}"
+    echo "  --cst: set cst binary path/filename   [default: ${CST_BINARY}]"
+    echo "  --csf-template: set CSF template file [default: ${CSF_TEMPLATE}]"
+    echo "  --spl: set filename for SPL binary    [default: ${SPL_FILE}]"
+    echo "  --key-dir: location for key files     [default: ${KEY_DIR}]"
+    echo "  --mfgtools: turn on clear / restore DCD addr for SPD SPL binary"
+    exit 0
+fi
+
+if [ "${DCD_CLEAR}" = "1" ]; then
+    MFGTOOLS="yes"
+else
+    MFGTOOLS="no"
+fi
+
+echo ""
+echo "SETTINGS FOR  : ${0}"
+echo "--------------:"
+echo "CST BINARY    : ${CST_BINARY}"
+echo "CSF TEMPLATE  : ${CSF_TEMPLATE}"
+echo "SPL FILENAME  : ${SPL_FILE}"
+echo "KEYS DIRECTORY: ${KEY_DIR}"
+echo "MFGTOOLS      : ${MFGTOOLS}"
+echo ""
+
+# Transform template -> config
+sed "s/@@KEY_ROOT@@/${KEY_DIR}/g" ${CSF_TEMPLATE} > ${CSF_TEMPLATE}.csf-config
+
+# working SPL file used for signature
+cp ${SPL_FILE} ${SPL_FILE}.mod
+
+# DCD address must be cleared for signature, as mfgtool will clear it.
+if [ "${DCD_CLEAR}" = "1" ]; then
+    # generate a NULL address for the DCD
+    dd if=/dev/zero of=zero.bin bs=1 count=4
+    # replace the DCD address with the NULL address
+    dd if=zero.bin of=${SPL_FILE}.mod seek=12 bs=1 conv=notrunc
+    rm zero.bin
+
+    # append DCD block information to CSF config
+    # HACK: Fix DCD block information below: replacing initial value with 0x2f010000
+    # when fixed use: cut -d' ' -f4-5
+    DCD_BLOCKS="0x2f010000 $(mkimage -l ${SPL_FILE} | grep 'DCD Blocks' | tr -s ' ' | cut -d' ' -f4-5)"
```

Comment:
Should we have an option here? Hardcoding for ulp will break imx6/7 (they use the hardcoded value provided by mkimage).

Reply:
Yep nice catch.  Not sure how we want to handle.  In my latest version of the script, I removed mkimage requirement entirely by using od to read hex values.  Let me see if one of the other fields has the base of RAM in it.

Reply:
So in both cases (m4 binary and SPL) the 0x20 offset from the IVT contains the base of RAM where I'm assuming the binary would be loaded.  I would need to see a binary from i.MX6/7 to confirm if it's there too.  If so we can use that value for the DCD fix.

Reply:
Yeah, I believe extracting from the IVT is the most generic way.

Reply:
Fixed in latest version